File: tank_player_templates/tank_player_cs/src/Player.cs
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
using System;
using System.Threading;
using Safir.Dob.Typesystem;

namespace tank_player_cs
{
	class Player : 	Safir.Dob.Dispatcher,
					Safir.Dob.StopHandler,
					Safir.Dob.EntityHandler,
					Safir.Dob.EntitySubscriber
	{
		private AutoResetEvent dispatchEvent=new AutoResetEvent(false);
		private bool running=true;
		private Safir.Dob.Connection connection=new Safir.Dob.Connection();
		private TankLogic logic=null;

		private HandlerId myHandlerId=null;
		private InstanceId currentGameId=null;
		private InstanceId myPlayerId=null;
		private InstanceId myJoystickId=null;
		private int myTankId=-1;
		private int joystickCounter=0;

		public Player ()
		{
			//This players unique identifiers
			myHandlerId = new HandlerId (TankLogic.PlayerName + "Handler");
			myPlayerId = new InstanceId (TankLogic.PlayerName + "Instance");

			// Open DOB connection. Register player and joystick, subscribe for gameStates
			connection.Open (TankLogic.PlayerName, "", 0, this, this);
			connection.RegisterEntityHandler (Consoden.TankGame.Player.ClassTypeId, myHandlerId, Safir.Dob.InstanceIdPolicy.Enumeration.HandlerDecidesInstanceId, this);
			connection.RegisterEntityHandler (Consoden.TankGame.Joystick.ClassTypeId, myHandlerId, Safir.Dob.InstanceIdPolicy.Enumeration.HandlerDecidesInstanceId, this);
			connection.SubscribeEntity (Consoden.TankGame.GameState.ClassTypeId, this);
	        
			//Create our player entity
			Consoden.TankGame.Player player = new Consoden.TankGame.Player ();
			player.Name.Val = TankLogic.PlayerName;	        
			connection.SetAll (player, myPlayerId, myHandlerId);

			//Run the game player
			Run ();
		}

		public void OnDoDispatch ()
		{
			dispatchEvent.Set (); //signal to main thread
		}

		public void OnStopOrder ()
		{
			running = false;
			dispatchEvent.Set ();
		}

		public void OnRevokedRegistration (long typeId, Safir.Dob.Typesystem.HandlerId handlerId)
		{
			System.Console.WriteLine("Revoked by handler "+handlerId.ToString());
		}

		public void OnCreateRequest (Safir.Dob.EntityRequestProxy entityRequestProxy, Safir.Dob.ResponseSender responseSender)
		{
			responseSender.Send (Safir.Dob.ErrorResponse.CreateErrorResponse (Safir.Dob.ResponseGeneralErrorCodes.SafirNoPermission, ""));
		}

		public void OnUpdateRequest (Safir.Dob.EntityRequestProxy entityRequestProxy, Safir.Dob.ResponseSender responseSender)
		{
			responseSender.Send (Safir.Dob.ErrorResponse.CreateErrorResponse (Safir.Dob.ResponseGeneralErrorCodes.SafirNoPermission, ""));
		}

		public void OnDeleteRequest (Safir.Dob.EntityRequestProxy entityRequestProxy, Safir.Dob.ResponseSender responseSender)
		{
			responseSender.Send (Safir.Dob.ErrorResponse.CreateErrorResponse (Safir.Dob.ResponseGeneralErrorCodes.SafirNoPermission, ""));
		}

		public void OnNewEntity (Safir.Dob.EntityProxy entityProxy)
		{
			Consoden.TankGame.GameState gameState = entityProxy.Entity as Consoden.TankGame.GameState;
			if (gameState == null) {
				return;
			}

			for (int i=0; i<gameState.Tanks.Count; i++) {
				if (!gameState.Tanks [i].IsNull ()) {
					Consoden.TankGame.Tank tank = gameState.Tanks [i].Obj;
					if (tank.PlayerId.Val == myPlayerId) {
						currentGameId = entityProxy.InstanceId;
						myTankId = tank.TankId.Val;
						myJoystickId = InstanceId.GenerateRandom ();

						Consoden.TankGame.Joystick joystick = new Consoden.TankGame.Joystick ();
						joystickCounter = 0;
						joystick.PlayerId.Val = myPlayerId;
						joystick.GameId.Val = currentGameId;
						joystick.TankId.Val = myTankId;
						joystick.Counter.Val = joystickCounter++;
						connection.SetAll (joystick, myJoystickId, myHandlerId);
						logic = new TankLogic (myTankId, UpdateJoystick);
						try
						{
						    logic.MakeMove (gameState);
					    }
					    catch
					    {
					        Console.WriteLine("Caught unhandled exception in TankLogic!");
					    }
						break;
					}
				}
			}
		}

		public void OnUpdatedEntity (Safir.Dob.EntityProxy entityProxy)
		{
			if (entityProxy.TypeId == Consoden.TankGame.GameState.ClassTypeId &&
				entityProxy.InstanceId == currentGameId && logic!=null) {
				Consoden.TankGame.GameState gs = (Consoden.TankGame.GameState)entityProxy.Entity;
				if (gs.Winner.Val == Consoden.TankGame.Winner.Enumeration.Unknown) {
				    try
					{
					    logic.MakeMove((Consoden.TankGame.GameState)entityProxy.Entity);
				    }
				    catch
				    {
				        Console.WriteLine("Caught unhandled exception in TankLogic!");
				    }
				}
			}
		}

		public void OnDeletedEntity (Safir.Dob.EntityProxy entityProxy, bool deprecated)
		{
			if (entityProxy.TypeId == Consoden.TankGame.GameState.ClassTypeId &&
				entityProxy.InstanceId == currentGameId) {
				if (myJoystickId != null) {
					connection.Delete (new EntityId (Consoden.TankGame.Joystick.ClassTypeId, myJoystickId), myHandlerId);
				}

				logic = null;
				myJoystickId = null;
				currentGameId = null;
				myTankId = -1;
			}
		}

		private void UpdateJoystick (Consoden.TankGame.Direction.Enumeration moveDirection,
		                            Consoden.TankGame.Direction.Enumeration towerDirection,
		                            bool fire, bool dropMine)
		{
			if (myJoystickId == null) {
				return; //we are not active in a game
			}

			Consoden.TankGame.Joystick joystick = new Consoden.TankGame.Joystick ();
			joystick.PlayerId.Val = myPlayerId;
			joystick.GameId.Val = currentGameId;
			joystick.TankId.Val = myTankId;
			joystick.Counter.Val = joystickCounter++;
			joystick.MoveDirection.Val = moveDirection;
			joystick.TowerDirection.Val = towerDirection;
			joystick.Fire.Val = fire;
			joystick.MineDrop.Val = dropMine;
			connection.SetAll (joystick, myJoystickId, myHandlerId);
		}

		private void Run ()
		{
			Console.WriteLine (TankLogic.PlayerName + " is running");
			while (running) {
				dispatchEvent.WaitOne ();
				connection.Dispatch();
			}
		}

		public static void Main (string[] args)
		{
			new Player();
		}
	}
}

File: tank_player_templates/tank_player_java/src/cwg/Player.java
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
package cwg;

import com.saabgroup.safir.dob.typesystem.*;
import com.saabgroup.safir.dob.ErrorResponse;
import com.saabgroup.safir.dob.ResponseGeneralErrorCodes;
import consoden.tankgame.Direction;

class Player implements com.saabgroup.safir.dob.Dispatcher,
                    com.saabgroup.safir.dob.StopHandler,
                    com.saabgroup.safir.dob.EntityHandler,
                    com.saabgroup.safir.dob.EntitySubscriber,
                    TankLogic.JoystickHandler {
    
	private boolean running=true;
    private com.saabgroup.safir.dob.Connection dobConnection;
    private boolean dispatchSignal=false;
    private InstanceId currentGameId = null;
    private HandlerId myHandlerId = null;
    private InstanceId myPlayerId = null;
    private InstanceId myJoystickId = null;
    private int myTankId = -1;
    private TankLogic logic = null;
    private int joystickCounter = 0;
        
    public Player() {
    	//set up our unique id's
        myHandlerId=new HandlerId(TankLogic.PLAYER_NAME+"Handler");
        myPlayerId= new InstanceId(TankLogic.PLAYER_NAME+"Instance");
        
        //open DOB connection. Register player and joystick. Subscribe for gameStates.
        dobConnection = new com.saabgroup.safir.dob.Connection();
        dobConnection.open(TankLogic.PLAYER_NAME, "", 0, this, this);
        dobConnection.registerEntityHandler(consoden.tankgame.Player.ClassTypeId, myHandlerId, com.saabgroup.safir.dob.InstanceIdPolicy.HANDLER_DECIDES_INSTANCE_ID, this);
        dobConnection.registerEntityHandler(consoden.tankgame.Joystick.ClassTypeId, myHandlerId, com.saabgroup.safir.dob.InstanceIdPolicy.HANDLER_DECIDES_INSTANCE_ID, this);
        dobConnection.subscribeEntity(consoden.tankgame.GameState.ClassTypeId, this);
        
        //create our player entity
        consoden.tankgame.Player player=new consoden.tankgame.Player();
        player.name().setVal(TankLogic.PLAYER_NAME);
        dobConnection.setAll(player, myPlayerId, myHandlerId);

        //run the game player
        run();
    }

    @Override
    public void onRevokedRegistration(long typeId, com.saabgroup.safir.dob.typesystem.HandlerId handlerId) {
    	
    	if (handlerId==myHandlerId) {
    		System.out.println("Revoked!");
    		onStopOrder();
    	}
    }
    
    @Override
    public void onCreateRequest(com.saabgroup.safir.dob.EntityRequestProxy entityRequestProxy, 
            com.saabgroup.safir.dob.ResponseSender responseSender) {
            
    	responseSender.send(ErrorResponse.createErrorResponse(ResponseGeneralErrorCodes.getSafirNoPermission(), ""));
    }

    @Override
    public void onDeleteRequest(com.saabgroup.safir.dob.EntityRequestProxy entityRequestProxy, 
            com.saabgroup.safir.dob.ResponseSender responseSender) {
    	
    	responseSender.send(ErrorResponse.createErrorResponse(ResponseGeneralErrorCodes.getSafirNoPermission(), ""));
    }

    @Override
    public void onUpdateRequest(com.saabgroup.safir.dob.EntityRequestProxy entityRequestProxy, 
            com.saabgroup.safir.dob.ResponseSender responseSender) {
            
    	responseSender.send(ErrorResponse.createErrorResponse(ResponseGeneralErrorCodes.getSafirNoPermission(), ""));
    }
    
    @Override
    public void onNewEntity(com.saabgroup.safir.dob.EntityProxy entityProxy) {

    	if (entityProxy.getEntity() instanceof consoden.tankgame.GameState) {
    			
    		consoden.tankgame.GameState gameState=(consoden.tankgame.GameState)entityProxy.getEntity();
    		
    		for (int i=0; i<gameState.tanks().size(); i++) {
    			if (!gameState.tanks().get(i).isNull()) {
    				consoden.tankgame.Tank tank=gameState.tanks().get(i).getObj();
    				if (tank.playerId().getVal().equals(myPlayerId)) {
    					this.currentGameId=entityProxy.getInstanceId();
    					this.myTankId=tank.tankId().getVal();
    					this.myJoystickId=InstanceId.generateRandom();
    					
    					consoden.tankgame.Joystick joystick=new consoden.tankgame.Joystick();
    					joystickCounter=0;
    					joystick.playerId().setVal(myPlayerId);
    					joystick.gameId().setVal(currentGameId);
    					joystick.tankId().setVal(myTankId);
    					joystick.counter().setVal(joystickCounter++);
    					dobConnection.setAll(joystick, myJoystickId, myHandlerId);
    					logic=new TankLogic(myTankId, this);
    					break;
    				}
    			}
    		}
    	}
    }
    
    @Override
    public void onUpdatedEntity(com.saabgroup.safir.dob.EntityProxy entityProxy) {
    	if (entityProxy.getEntity() instanceof consoden.tankgame.GameState &&
    			entityProxy.getInstanceId().equals(currentGameId) && logic!=null) {
            
            consoden.tankgame.GameState gs = (consoden.tankgame.GameState)entityProxy.getEntity();
            if (gs.winner().getVal() == consoden.tankgame.Winner.UNKNOWN) {
                try {
				    logic.makeMove((consoden.tankgame.GameState)entityProxy.getEntity());
			    } 
			    catch (java.lang.Exception e) {
				    System.out.println("Caught unhandled exception in TankLogic!");
			    }
                
            }
    	}
    }
    
    @Override
    public void onDeletedEntity(com.saabgroup.safir.dob.EntityProxy entityProxy, boolean deprecated) {
    	if (entityProxy.getTypeId()==consoden.tankgame.GameState.ClassTypeId && 
    			entityProxy.getInstanceId().equals(currentGameId)) {
    		if (myJoystickId!=null) {
    			dobConnection.delete(new EntityId(consoden.tankgame.Joystick.ClassTypeId, myJoystickId), myHandlerId);
    		}
    		
    		logic=null;            
            myJoystickId=null;
            currentGameId=null;
            myTankId=-1;
        }
    }

    @Override
    public void onStopOrder() {
    	synchronized(this) {
            running=false;
            this.notify();
        }
    }
    
    @Override
    public void onDoDispatch() {
        synchronized(this) {
            dispatchSignal=true;
            this.notify();
        }
    }
    
    private void run() {
    	System.out.println(TankLogic.PLAYER_NAME+" is running");
    	while (running) {
	        synchronized(this) {
	            while(!dispatchSignal){
	            	if (!running)
	            		return;
	            	
	                try {
	                    this.wait();
	                } catch(InterruptedException e){}
	            }

	            dispatchSignal=false;
	            dobConnection.dispatch();
	        }
    	}
    }
    
	@Override
	public void setJoystick(Direction moveDirection, Direction towerDirection, boolean fire, boolean dropMine) {

	    if (myJoystickId==null) {
	        return; //we are not active in a game
	    }
	    
	    if (moveDirection==null)
	        moveDirection=consoden.tankgame.Direction.NEUTRAL;
        if (towerDirection==null)
	        towerDirection=consoden.tankgame.Direction.NEUTRAL;

	    consoden.tankgame.Joystick joystick=new consoden.tankgame.Joystick();
		joystick.playerId().setVal(myPlayerId);
		joystick.gameId().setVal(currentGameId);
		joystick.tankId().setVal(myTankId);
		joystick.counter().setVal(joystickCounter++);
		joystick.moveDirection().setVal(moveDirection);
		joystick.towerDirection().setVal(towerDirection);
		joystick.fire().setVal(fire);
		joystick.mineDrop().setVal(dropMine);
		dobConnection.setAll(joystick, myJoystickId, myHandlerId);
	}
	
	public static void main(String[] args) {
        new Player();
    }
}
